File: Data Structures & Algorithms/valid-sudoku/submission-1.cs
public class Solution {
    public bool IsValidSudoku(char[][] board) {
        int nrows = board.Length;
        int ncols = board[0].Length;
        Dictionary<int, HashSet<char>> seenInRow = new();
        Dictionary<int, HashSet<char>> seenInCol = new();
        Dictionary<(int, int), HashSet<char>> seenInSquare = new();
        for (int row = 0; row < nrows; row++) {
            for (int col = 0; col < ncols; col++) {
                char val = board[row][col];
                if (val == '.') {
                    continue;
                }
                if (!SeenInRow(val, seenInRow, row)){
                    return false;
                }
                if (!SeenInCol(val, seenInCol, col)) {
                    return false;
                }
                if (!SeenInSquare(val, seenInSquare, row, col)) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    public bool SeenInSquare(char value, Dictionary<(int, int), HashSet<char>> seenInSquare, int row, int col) {
        var squareId = (row / 3, col / 3);
        if (!seenInSquare.TryGetValue(squareId, out var seen)) {
            seen = new();
            seenInSquare[squareId] = seen;
        }
        
        return seen.Add(value);
    }
    
    public bool SeenInRow(char value, Dictionary<int, HashSet<char>> seenInRow, int row) {
        if (!seenInRow.TryGetValue(row, out var seen))
        {
            seen = new();
            seenInRow[row] = seen;
        }

        return seen.Add(value);
    }
    
    public bool SeenInCol(char value, Dictionary<int, HashSet<char>> seenInCol, int col) {
        
        if (!seenInCol.TryGetValue(col, out var seen))
        {
            seen = new();
            seenInCol[col] = seen;
        }
        
        return seen.Add(value);
    }
}

File: Data Structures & Algorithms/search-for-word/submission-1.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        nrows, ncols = len(board), len(board[0])
        seen = set()
        directions = [(0, -1), (0, 1), (-1, 0), (1, 0)]

        def search(row, col, index):            
            if board[row][col] != word[index]:
                return False

            seen.add((row, col))

            if index == len(word) - 1:
                return True


            for dr, dc in directions:
                nr, nc = row + dr, col + dc
                if (nr, nc) in seen or nr < 0 or nr >= nrows or nc < 0 or nc >= ncols:
                    continue
                
                value = search(nr, nc, index + 1)
                if value:
                    return True

            seen.remove((row, col))
            return False

        for row in range(nrows):
            for col in range(ncols):
                if search(row, col, 0):
                    return True
        return False

